hera_librarian/async_transfers/globus.py

```diff
@@ -0,0 +1,271 @@
+        # Globus transfer.
+        if local_path.is_dir():
+            transfer_data.add_item(str(local_path), str(remote_path), recursive=True)
+        else:
+            transfer_data.add_item(str(local_path), str(remote_path), recursive=False)
+
+        # try to submit the task
+        try:
+            task_doc = transfer_client.submit_transfer(transfer_data)
+        except globus_sdk.TransferAPIError as e:
+            return False
+
+        self.task_id = task_doc["task_id"]
+        return True
+
+    def batch_transfer(
+        self,
+        paths: list[tuple[Path]],
+        local_endpoint,
+        remote_endpoint,
+    ) -> bool:
+        self.transfer_attempted = True
+
+        # We have to do a lot of the same legwork as above for a single
+        # transfer, with the biggest change being that we can add multiple items
+        # to a single TransferData object. This is effectively how we "batch" books
+        # using Globus.
+
+        # start by authorizing
+        if not self.authorize():
+            return False
+
+        # make a label from the first book
+        label = "batch with " + os.path.basename(paths[0][0])
+
+        # create a transfer client to handle the transfer
+        transfer_client = globus_sdk.TransferClient(authorizer=self.authorizer)
+
+        # get a TransferData object
+        transfer_data = self._get_transfer_data(local_endpoint, remote_endpoint, label)
+
+        # add each of our books to our task
+        for local_path, remote_path in paths:
+            # We need to figure out if the local path is actually a directory or a
+            # flat file, which annoyingly requires different handling as part of the
+            # Globus transfer.
+            if local_path.is_dir():
+                transfer_data.add_item(
+                    str(local_path), str(remote_path), recursive=True
+                )
+            else:
+                transfer_data.add_item(
+                    str(local_path), str(remote_path), recursive=True
+                )
+
+        # submit the transfer
+        try:
+            task_doc = transfer_client.submit_transfer(transfer_data)
+        except globus_sdk.TransferAPIError as e:
+            return False
+
+        self.task_id = task_doc["task_id"]
+        return True
+
+    @property
+    def transfer_status(self) -> TransferStatus:
+        """
+        Query Globus to see if our transfer has finished yet.
+        """
+        if not self.authorize():
+            # We *should* be able to just assume that we have already
+            # authenticated and should be able to query the status of our
+            # transfer. However, if for whatever reason we're not able to talk
+            # to Globus (network issues, Globus outage, etc.), we won't be able
+            # to find out our transfer's status -- let's bail and assume we
+            # failed
+            return TransferStatus.FAILED
```

Comment:
This is a little dangerous... Maybe we should consider having an exception be raised here and catch it in the next layer above